pypred/ast.py

```diff
@@ -307,38 +307,96 @@ def failure_info(self, ctx):
 
 
 class ContainsOperator(Node):
-    "Used for the 'contains' operator"
-    def __init__(self, left, right):
+    "Used for the 'contains' operator and derived list comparisons"
+
+    def __init__(self, op, left, right):
+        self.type = op
         self.left = left
         self.right = right
 
+    def name(self):
+        return "%s operator at %s" % (self.type.upper(), self.position)
+
     def _validate(self, info):
-        if not isinstance(self.right, (Number, Literal, Constant)):
+        if self.type not in ("contains", "any", "all"):
             errs = info["errors"]
-            errs.append("Contains operator must take a literal or constant! Got: %s for %s" % (repr(self.right), self.name()))
+            errs.append("Unknown compare operator %s" % self.type)
+            return False
+
+        if self.type == "contains" and not isinstance(
+            self.right, (Number, Literal, Constant)
+        ):
+            errs = info["errors"]
+            errs.append(
+                "Contains operator must take a literal or constant! Got: %s for %s"
+                % (repr(self.right), self.name())
+            )
             return False
+
+        if self.type in ("any", "all") and not isinstance(self.right, LiteralSet):
+            errs = info["errors"]
+            errs.append(
+                "Contains %s operator must take a set! Got: %s for %s"
+                % (self.type, repr(self.right), self.name())
+            )
+            return False
+
         return True
 
     @failure_info
     def eval(self, ctx):
         left = self.left.eval(ctx)
-        if hasattr(left, "__contains__"):
-            right = self.right.eval(ctx)
+        if not hasattr(left, "__contains__"):
+            return False
+
+        right = self.right.eval(ctx)
+        if self.type == "contains":
             return right in left
+
+        if not hasattr(right, "__iter__"):
```

Comment:
Why is this not needed for `left`? What makes this check for left?

Reply:
It's because we don't iterate on `left`, so we just check that `hasattr(left, "__contains__")` since this is the method that gets executed in `x in y` IIUC

Reply:
I'm not sure I follow, we do iterate over both `left` and `right` in [this expression](https://github.com/dialoguemd/pypred/pull/2/files#diff-ff6c318c7d52ada0696e22d0127c9cd6e215e90e9415d173bf1f6a4835876faaR359)

Reply:
Actually, the way I understand it is that when you do `x in y` in python, the `__contains__` method gets executed. So for the expression `any(right_item in left for right_item in right)`, we iterate over `right`, for each items we check that `right_item in left` and the `any` function exits as soon as there is a `True` returned. So we never call the `__iter__` function on the `left` parameter

Reply:
Ah yes my bad, makes sense. `for in` and `in` are obviously different operators.